### ISSUE TITLE
[Enhancement]: job_stats informational view is not updated when executed manually

### ISSUE THREAD

**Issue by henriquegelio** [timescale/timescaledb#5188]
### What type of enhancement is this?

API improvement

### What subsystems and features will be improved?

User-Defined Action (UDA)

### What does the enhancement do?

Hey folks,

We're currently using timescaledb as storage for a time-event based application.

Many events arrives every second and we're storing them into timescale to perform some general purpose aggregations and, obviously, to compress data more efficiently. Basically, an event is a jsonb structure with a separate timestamp column (in milliseconds).

In our application, every "event type" is a different hypertable in our model, and an admin user can decide if he wants to compress it manually, define a compression policy, and even compress manually if there is already a compression policy defined.

Think of this use case: we've imported a huge amount of data (with hundreds of GB) from another database into timescaledb. The user wants to create a policy in which all **new** events will get compressed after 6 months, but everything that is already imported should be instantly compressed. So he will define a policy to compress events after 6 months old, and manually compress the whole hypertable.
Basically this is the use case we're trying to achieve.
By manually compress I mean an API that does the `SELECT compress_chunk(i, true) FROM show_chunks(hypertable) AS i`.

The problem is that as our hypertable might have hundreds of GB, thus a synchronous call to  `SELECT compress_chunk(i, true) FROM show_chunks(hypertable) AS i` is not wanted due to the time the operation may take. We've thought about it being asynchronous.

So here's what we tried:

- Define a user defined action that receives a configuration with the hypertable name and schema that should be compressed, and manually compress the hypertable with `SELECT compress_chunk(i, true) FROM show_chunks(hypertable) AS i` inside the UDA
- Register this UDA with `add_job()` (as we want this job to be run only ONCE, we've set scheduled to false and next_start to now())

Unfortunately, this approach didn't work as expected because, obviously, as scheduled is false, the first run never happens.

Our next try was to force run the job with `run_job()`, passing the job_id returned by `add_job()`.
Unfortunately, this workaround also don't work as we expect because of two reasons:

1. When a job is run with `run_job()`, the informational view `job_stats` is not updated, probably because it's run in foreground.
2. The statement `CALL run_job(id)` is synchronous so it does not serve our purpose

Well, the 2nd reason above could be dealt by triggering a thread in our application backend that calls `run_job()` and don't wait for it to end. So our main goal would be to consume the informational view and check if the job succeeded or not.
But, for the 1st reason, we couldn't see a workaround.

So here goes my first main question:
Is this behaviour (not updating the informational view) expected when a job is run in foreground? We understand about the workers and so, but it would be really nice to have the views being updated not only in background by the workers, but every time a job is run (foreground as well).
Maybe our specific async use case is not the expected usage of this automation, but even for sync operations it would be nice to have the views updated by `run_job()`.

The other question is regarding a possible workaround we've thought: Instead of setting scheduled to false, it would remain true but with a scheduled_interval very spaced, for example 10 years. So after the first run of this job is done, either with errors or not, we could pause the job by setting the scheduled to false again with `alter_job()` or simply deleting it with `delete_job()`. However, this workaround seems a little unusual (ugly).

With that said, here goes the second question: Is there a smarter way to implement the use case we intend to using timescaledb, or the above workaround is the only way?

Thanks in advance for the help and feel free to ask for additional information if relevant.

### Implementation challenges

The purpose is to have timescaledb_information.job_stats updated even if a job is run in foreground

**Comment by konskov** [comment]
hi @henriquegelio thank you for reaching out. 

To answer your first question: Yes, this behavior is expected, as the view is only concerned with background jobs. So it only logs the executions of background jobs. Calling `run_job` does not execute the job in the background by a worker, but by the session backend instead, so the view is not updated, as you also point out. You would have to keep track of the foreground executions yourself. 

For the second question: The workaround you are using is what I would also suggest. (make sure the job runs once and then unschedule it or delete it). 
An alternative of the sort `PERFORM alter_job(jobid, scheduled => false);` at the end of the UDA might work but I would not recommend it because it has led to locking issues in the past. 
Currently our jobs API does not support submitting a task for a single async execution but rather for repeated executions on a schedule, so you will have to use the workaround to achieve that for the time being. However, it is on our roadmap to add support for this particular use case.

**Comment by henriquegelio** [comment]
Hi @konskov ,

Thanks for you answer. We're doing as suggested (setting schedule to false) after the job completes.

It would be nice to have such a feature to run the jobs only once instead of this workaround!
I'll keep track of the recent releases in case it gets implemented.

Feel free to close this issue if desired
